src/libraries/System.Console/src/System/Console.cs

```diff
@@ -22,7 +22,7 @@ public static class Console
         // there's little benefit to having a large buffer.  So we use a smaller buffer size to reduce working set.
         private const int WriteBufferSize = 256;
 
-        private static object InternalSyncObject = new object(); // for synchronizing changing of Console's static fields
+        internal static object InternalSyncObject = new object(); // for synchronizing changing of Console's static fields
```

Comment:
What remains to be using this?  To avoid further issues, it'd be good if this could be relegated to only being used for things unrelated to input/output, e.g. if it was only for the cancel callbacks (and renamed accordingly).  Is that possible?

Reply:
The `Monitor.IsEntered` check in `ConsolePal.Unix` still checks for this.
Should I remove the check? It guards us from potential deadlocks involving `InternalSyncObject` and `Console.Out`.

Reply:
> relegated to only being used for things unrelated to input/output, e.g. if it was only for the cancel callbacks (and renamed accordingly). Is that possible?

I'll take a look.

Reply:
I've reduced the usage of the lock to the event handlers. I think the usage of `Interlocked`/`Volatile` is proper to ensure thread-safety.